compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/JvmInlineClassLowering.kt

```diff
@@ -0,0 +1,402 @@
+                    }
+                }
+
+                function.body = irExprBody(call)
+            }
+        }
+
+        return listOf(worker, function)
+    }
+
+    // Secondary constructors for boxed types get translated to static functions returning
+    // unboxed arguments. We remove the original constructor.
+    private fun transformConstructorFlat(constructor: IrConstructor, replacement: IrReplacementFunction): List<IrDeclaration> {
+        val worker = replacement.function
+
+        scoped {
+            addMappings(replacement.valueParameterMap)
+            worker.valueParameters.forEach { it.transformChildrenVoid() }
+            worker.body = context.createIrBuilder(worker.symbol).irBlockBody(worker) {
+                val thisVar = irTemporaryVarDeclaration(
+                    worker.returnType, nameHint = "\$this", isMutable = false
+                )
+                addMapping(constructor.constructedClass.thisReceiver!!.symbol, thisVar)
+
+                constructor.body?.contents?.forEach { statement ->
+                    +statement
+                        .transform(object : IrElementTransformerVoid() {
+                            // Don't recurse under nested class declarations
+                            override fun visitClass(declaration: IrClass): IrStatement {
+                                return declaration
+                            }
+
+                            // Capture the result of a delegating constructor call in a temporary variable "thisVar".
+                            //
+                            // Within the constructor we replace references to "this" with references to "thisVar".
+                            // This is safe, since the delegating constructor call precedes all references to "this".
+                            override fun visitDelegatingConstructorCall(expression: IrDelegatingConstructorCall): IrExpression {
+                                expression.transformChildrenVoid()
+                                return irSetVar(thisVar.symbol, expression)
+                            }
+
+                            // A constructor body has type unit and may contain explicit return statements.
+                            // These early returns may have side-effects however, so we still have to evaluate
+                            // the return expression. Afterwards we return "thisVar".
+                            // For example, the following is a valid inline class declaration.
+                            //
+                            //     inline class Foo(val x: String) {
+                            //       constructor(y: Int) : this(y.toString()) {
+                            //         if (y == 0) return throw java.lang.IllegalArgumentException()
+                            //         if (y == 1) return
+                            //         return Unit
+                            //       }
+                            //     }
```

Comment:
Please ensure there's a test, or add a new one, that checks side effects here

Reply:
Already done! I've added `testSimpleSecondaryConstructor` in the inline class tests which checks this case.